crates/common/src/l2.rs

```diff
@@ -57,6 +60,61 @@ enum SyncError {
     Unknown(String),
 }
 
+struct SequentialL2BlockBuffer {
+    blocks: BTreeMap<u64, L2BlockResponse>,
+    next_expected_height: u64,
+    notify: Arc<Notify>,
+}
+
+impl SequentialL2BlockBuffer {
+    fn new(start_height: u64) -> Self {
+        Self {
+            blocks: BTreeMap::new(),
+            next_expected_height: start_height,
+            notify: Arc::new(Notify::new()),
+        }
+    }
+
+    fn add_block(&mut self, block: L2BlockResponse) {
+        let height = block.header.height.to();
+
+        if height < self.next_expected_height || self.blocks.contains_key(&height) {
+            return;
+        }
+
+        self.blocks.insert(height, block);
+        self.notify.notify_one();
+    }
+
+    fn extend_blocks(&mut self, mut blocks: BTreeMap<u64, L2BlockResponse>) {
+        blocks.retain(|&k, _| k >= self.next_expected_height);
+        let should_notify = !blocks.is_empty();
+
+        self.blocks.extend(blocks);
+
+        if should_notify {
+            self.notify.notify_one();
+        }
+    }
+
+    /// Drain from next_expected_height until the first gap or until the buffer is fully consumed
+    /// Returns an ordered vector of `L2BlockResponse`
+    fn drain_sequential(&mut self) -> Vec<L2BlockResponse> {
+        let mut sequential_blocks = Vec::new();
+
+        while let Some(block) = self.blocks.remove(&self.next_expected_height) {
```

Comment:
I don't know if we have a 100% guarantee that there will be NO gaps when we `.add_block`. But if there are, then it's gonna be a big problem.